helix-term/src/commands.rs

```diff
@@ -2410,17 +2410,39 @@ impl ui::menu::Item for MappableCommand {
     fn label(&self, keymap: &Self::Data) -> Spans {
         // formats key bindings, multiple bindings are comma separated,
         // individual key presses are joined with `+`
+        //
+        // This closure does as few allocations as possible by using the `Display` impl of
+        // `KeyEvent`s and not using `slice::join`, which needs an intermediate `Vec` to be
+        // allocated for each sequence of keys.
         let fmt_binding = |bindings: &Vec<Vec<KeyEvent>>| -> String {
-            bindings
-                .iter()
-                .map(|bind| {
-                    bind.iter()
-                        .map(|key| key.to_string())
-                        .collect::<Vec<String>>()
-                        .join("+")
-                })
-                .collect::<Vec<String>>()
-                .join(", ")
+            const OOM: &str = "Writing to a string can only fail on an Out-Of-Memory error";
+
+            let mut res = String::new();
+
+            fn write_binding(res: &mut String, binds: &[KeyEvent]) {
```

Comment:
Thanks for your optimization PRs! I'm glad to see contributors helping improve performance where possible.

Although, in my opinion, this particular change is a regression in readability. The original code is pretty clear and concise in its use of combinators, whereas this requires more than a quick glance to understand what it's doing.

I applaud optimizations in general, but personally I think readability is more important if performance is not suffering. And this code is not in any kind of hot path as far as I can tell.

Reply:
Yes I could remove this one, it's for the command palette, not as big of a deal as the one for LSP references. I didn't check if helix already has `itertools` in its dependencies, I may be able to avoid allocations without sacrificing readability that way

Reply:
Made a much clearer version using `fold`